### ISSUE TITLE
Accessible file tree

### ISSUE THREAD

**Issue by felixfbecker** [sourcegraph/sourcegraph#12916]
## Todo

- [x] Handle errors (i.e when the list of items is truncated) 
- [x] What to do with the `..` case? https://sourcegraph.slack.com/archives/C04931KQVRC/p1673961506582769
  - [x] Implement the old behavior behind a flag? https://github.com/sourcegraph/sourcegraph/pull/46893
- [x] Handle submodules (and other features that I forgot?)
- [x] Create a Wildcard Tree
- [x] Migrate symbols to the new tree (need to look into the search issue https://github.com/dgreene1/react-accessible-treeview/issues/73)
- [x] expand single-child-folders

---

The file tree can be navigated with the arrow keys. However, this doesn't actually focus the DOM elements, it just gives them different colors and changes application state.

This means that when using a screen reader, the current element stays on the old item and the screen reader will not announce the selected items. Hitting tab will select a different element than expected.

![image](https://user-images.githubusercontent.com/10532611/89920900-4c356d00-dbfd-11ea-9fc3-79c8f060bee7.png)

Selecting an item with <kbd>enter</kbd> makes the search box steal focus:

![image](https://user-images.githubusercontent.com/10532611/89920992-6c652c00-dbfd-11ea-8838-29c2e56bdec5.png)

Additional comment:
@umpox We prefetch file data onMouseOver on the list, but we only support the same functionality for keyboard focusing in componentDidUpdate. We should be able to use onFocus here.

cc @attfarhan

**Comment by jasongornall** [comment]
@umpox this one still seems to be occurring added fixing label

**Comment by umpox** [comment]
Assigning myself to triage and check this has been covered in code intel audits and any other issues

**Comment by felixfbecker** [comment]
This is still an issue and actually got worse, because not even <kbd>tab</kbd> works anymore. The file tree is completely inoperable with the keyboard.

ARIA has a role/widget specification `tree`, which this should probably be. There may be an open source implementation we could use instead of our own, or we could try to mark up our tree with it. The same would need to be applied to the hierarchical symbols sidebar.

**Comment by lrhacker** [comment]
Additional notes:

- There may be some open source libraries we could consider bringing in to help. They would still need to be evaluated for things like bundle size, if the API is the best fit for our needs, outstanding issues they might have, etc., but here are a couple that seem promising:
  - [react-accessible-treeview](https://github.com/dgreene1/react-accessible-treeview)
  - [MUI Treeview](https://mui.com/material-ui/react-tree-view/)
- There is also work needing to be done on this component in relation to the scale audit, see #43657

**Comment by felixfbecker** [comment]
For this week before the audit, we shouldn't do anything more than a timeboxed removal of `tabindex="-1"` on the links as a slight mitigation (not the best though because it introduces a lot of tabstops). For a full fix, we need to replace the file tree component, which is going to take more time.

**Comment by philipp-spiess** [comment]
@taras-yemets and me looked into this a bit today (async). We looked at a few open source libraries that implement the [WAI tree view](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/) example. Particularly https://github.com/dgreene1/react-accessible-treeview looked very exciting since it allows us to manage the state ourself and has built-in helpers for async loading.

We think that moving our existing tree view to something like this could be a great improvement for our customers (better keyboard navigation, better accessibility, loading states are better communicated, etc).

There is one feature that our current tree view has that might be a bit harder to implement: Right now, when you do the initial load inside a sub-directory, (e.g. [this link](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/client/vscode)), the file tree only loads the current directory with a `..` on top to navigate out of it. This is a great feature since it allows large monorepo to behave more similarly to a multi-repo setup but it's quite confusing right now as it does a full reload of the tree view anytime you navigate out.

We have thought about a few approaches to support this in a new implementation:

1. We keep doing what we do. The initial render only lists the current dir and appends a `..` option (marked as a foldable). When this is selected, it will load the parent directory. We need a bit of patching then to process the tree structure properly and we need validation that we can properly move the focus afterwards (since we'll be deleting the `..` folder and need to move the focus somewhere else).
  
   <img width="287" alt="Screenshot 2023-01-12 at 17 16 34" src="https://user-images.githubusercontent.com/458591/212121762-736508e3-15a9-4b7c-870e-26b1f6135460.png">

2. We could add a new endpoint that loads all files inside a given tree and includes all parent-trees and their files for a more IDE like experience. E.g. for a link to `/client/vscode/` this would load `/`, `/client`, and `/client/vscode`. I think @felixfbecker mentioned in our team call on Tuesday that this is the behavior that we used to have in the past and we deliberately decided against it. Did I understand that correctly?
    
    ![image](https://user-images.githubusercontent.com/458591/212121890-bcb2dd92-fd51-4e45-bc23-52339441edad.png)

Does someone have any other ideas?

Another thing that we investigated are repos with large file trees that we can use to test this against, so I want to mention them here as well for easier access when testing:
- https://sourcegraph.com/github.com/DefinitelyTyped/DefinitelyTyped/-/tree/types
- https://sourcegraph.com/github.com/sgtest/megarepo/-/tree/vscode/src/vs/workbench/contrib/files/browser

And another consideration that we need to have is pagination. Here's what we currently show when we load over 2000 or 3000 directories:

![image](https://user-images.githubusercontent.com/458591/212122292-b6d4a327-1b01-48f4-9c4c-6bfbc7234668.png)

cc @sourcegraph/code-exploration

**Comment by philipp-spiess** [comment]
Some feedback from @vovakulikov and @felixfbecker: https://sourcegraph.slack.com/archives/C03CSAER9LK/p1673574976309089
